File: core/domain/cards/subject.go
package cards

import (
	"time"

	"github.com/google/uuid"
	"github.com/sixels/manekani/core/domain/cards/filters"
)

type (
	Subject struct {
		ID        uuid.UUID `json:"id"`
		CreatedAt time.Time `json:"created_at"`
		UpdatedAt time.Time `json:"updated_at"`

		Kind  string `json:"kind"`
		Level int32  `json:"level"`

		// e.g. "ground", "一", nil, "一", 2
		Name       string  `json:"name"`
		Value      *string `json:"value"`
		ValueImage *string `json:"value_image"`
		Slug       string  `json:"slug"`
		Priority   uint8   `json:"priority"`

		Resources           []Resource      `json:"resources"`
		StudyData           []StudyData     `json:"study_data"`
		AdditionalStudyData *map[string]any `json:"additional_study_data"`

		Dependencies []uuid.UUID `json:"dependencies"`
		Dependents   []uuid.UUID `json:"dependents"`
		Similars     []uuid.UUID `json:"similars"`
		Deck         uuid.UUID   `json:"deck"`

		Owner string `json:"owner"`
	}

	PartialSubject struct {
		ID uuid.UUID `json:"id"`

		Kind  string `json:"kind"`
		Level int32  `json:"level"`

		Name       string  `json:"name"`
		Value      *string `json:"value"`
		ValueImage *string `json:"value_image"`
		Slug       string  `json:"slug"`
		Priority   uint8   `json:"priority"`

		StudyData []StudyData `json:"study_data"`

		Dependencies []uuid.UUID `json:"dependencies"`
		Dependents   []uuid.UUID `json:"dependents"`
		Similars     []uuid.UUID `json:"similars"`
		Deck         uuid.UUID   `json:"deck"`

		Owner string `json:"owner"`
	}

	MinimalSubject struct {
		ID uuid.UUID `json:"id"`

		Kind  string `json:"kind"`
		Level int32  `json:"level"`

		Name       string  `json:"name"`
		Value      *string `json:"value"`
		ValueImage *string `json:"value_image"`
		Slug       string  `json:"slug"`
		Priority   uint8   `json:"priority"`

		Deck uuid.UUID `json:"deck"`
	}

	CreateSubjectRequest struct {
		Kind  string `json:"kind,omitempty" form:"kind" binding:"required"`
		Level int32  `json:"level,omitempty" form:"level" binding:"required"`

		Name       string  `json:"name" form:"name" binding:"required"`
		Value      *string `json:"value,omitempty" form:"value" binding:"-"`
		ValueImage *string `json:"value_image,omitempty" form:"value_image" binding:"-" swaggerignore:"true"`
		Slug       string  `json:"slug" form:"slug" binding:"required"`
		Priority   uint8   `json:"priority" form:"priority" binding:"required"`

		StudyData           []StudyData     `json:"study_data,omitempty" form:"study_data" binding:"-"`
		Resources           *[]Resource     `json:"resources,omitempty" form:"resources" binding:"-" swaggerignore:"true"`
		AdditionalStudyData *map[string]any `json:"additional_study_data,omitempty" form:"additional_study_data" binding:"-"`

		Dependencies []uuid.UUID `json:"dependencies,omitempty" form:"dependencies" binding:"-"`
		Dependents   []uuid.UUID `json:"dependents,omitempty" form:"dependents" binding:"-"`
		Similars     []uuid.UUID `json:"similars,omitempty" form:"similars" binding:"-"`

		Deck uuid.UUID `json:"deck" form:"deck" binding:"required"`
	}

	UpdateSubjectRequest struct {
		Kind  *string `json:"kind,omitempty" form:"kind"`
		Level *int32  `json:"level,omitempty" form:"level"`

		Name       *string `json:"name,omitempty" form:"name"`
		Value      *string `json:"value,omitempty" form:"value"`
		ValueImage *string `json:"value_image,omitempty" form:"value_image"`
		Slug       *string `json:"slug,omitempty" form:"slug"`
		Priority   *uint8  `json:"priority,omitempty" form:"priority"`

		StudyData           *[]StudyData    `json:"study_data,omitempty" form:"study_data"`
		Resources           *[]Resource     `json:"resources,omitempty" form:"resources"`
		AdditionalStudyData *map[string]any `json:"additional_study_data" form:"additional_study_data" binding:"-"`

		Dependencies *[]uuid.UUID `json:"dependencies,omitempty" form:"dependencies"`
		Dependents   *[]uuid.UUID `json:"dependents,omitempty" form:"dependents"`
		Similars     *[]uuid.UUID `json:"similars,omitempty" form:"similars"`
	}

	QueryManySubjectsRequest struct {
		filters.FilterPagination
		filters.FilterIDs
		filters.FilterKinds
		filters.FilterLevels
		filters.FilterSlugs
		filters.FilterDecks
		filters.FilterOwners
	}
)

type (
	Resource struct {
		URL      string            `json:"url"`
		Metadata map[string]string `json:"metadata"`
	}
	StudyData struct {
		Kind     string      `json:"kind"`
		Items    []StudyItem `json:"items"`
		Mnemonic string      `json:"mnemonic,omitempty"`
	}

	StudyItem struct {
		Value         string  `json:"value"`
		IsPrimary     bool    `json:"is_primary"`
		IsValidAnswer bool    `json:"is_valid_answer"`
		IsHidden      bool    `json:"is_hidden"`
		Category      *string `json:"category"`
		Resource      *string `json:"resource,omitempty"`
	}
)

File: api/src/graphql/subject.ts
import {
  CreateSubjectDto,
  Resource as DResource,
  StudyData as DStudyData,
  StudyDataItem as DStudyDataItem,
  Subject as DSubject,
  UpdateSubjectDto,
} from 'manekani-core';
import {
  Field,
  InputType,
  Int,
  ObjectType,
  OmitType,
  PartialType,
  PickType,
} from '@nestjs/graphql';

import { GraphQLJSONObject } from 'graphql-type-json';

@ObjectType()
export class Subject implements DSubject {
  @Field()
  readonly id: string;

  @Field()
  readonly createdAt: Date;
  @Field()
  readonly updatedAt: Date;

  @Field()
  readonly category: string;
  @Field(() => Int)
  readonly level: number;
  @Field()
  readonly name: string;
  @Field()
  readonly slug: string;
  @Field(() => Int)
  readonly priority: number;

  @Field({ nullable: true })
  readonly value?: string;
  @Field({ nullable: true })
  readonly valueImage?: string;

  @Field(() => [StudyData], { defaultValue: [] })
  readonly studyData: StudyData[];
  @Field(() => [Resource], { defaultValue: [] })
  readonly resources: Resource[];

  @Field(() => [String], { defaultValue: [] })
  readonly dependencies: string[];
  @Field(() => [String], { defaultValue: [] })
  readonly dependents: string[];
  @Field(() => [String], { defaultValue: [] })
  readonly similar: string[];

  @Field()
  readonly deckId: string;
  @Field()
  readonly ownerId: string;
}

@ObjectType()
@InputType('StudyDataItemInput')
export class StudyDataItem implements DStudyDataItem {
  @Field()
  readonly value: string;

  @Field()
  readonly isPrimary: boolean;

  @Field()
  readonly isValidAnswer: boolean;

  @Field()
  readonly isHidden: boolean;

  @Field({ nullable: true })
  readonly category?: string;

  @Field({ nullable: true })
  readonly resourceId?: string;
}

@ObjectType()
@InputType('StudyDataInput')
export class StudyData implements DStudyData {
  @Field()
  readonly type: string;

  @Field({ nullable: true })
  readonly mnemonic?: string;

  @Field(() => [StudyDataItem], { defaultValue: [] })
  readonly items: StudyDataItem[];
}

@ObjectType()
@InputType('ResourceInput')
export class Resource implements DResource {
  @Field()
  readonly name: string;

  @Field(() => GraphQLJSONObject)
  readonly metadata: Record<string, any>;
}

@InputType()
export class CreateSubjectInput
  extends PickType(
    Subject,
    [
      'category',
      'level',
      'name',
      'value',
      'valueImage',
      'slug',
      'priority',
      'dependencies',
      'dependents',
      'similar',
      'studyData',
      'resources',
    ] as const,
    InputType,
  )
  implements CreateSubjectDto {}

@InputType()
export class UpdateSubjectInput
  extends PartialType(CreateSubjectInput)
  implements UpdateSubjectDto {}
